### ISSUE TITLE
CSS transitions do not work well when using temporary_assigns

### ISSUE THREAD

**Issue by sn3p** [phoenixframework/phoenix_live_view#1757]
### Environment

* Elixir version (elixir -v): **1.12.3**
* Phoenix version (mix deps): **1.6.2**
* Phoenix LiveView version (mix deps): **1.17.5**
* Operating system: **macOS Big Sur (11.6)**
* Browsers you attempted to reproduce this bug on (the more the merrier): **Safari, Chome, Firefox, Edge, Opera**
* Does the problem persist after removing "assets/node_modules" and trying again? Yes/no: **Yes**

### Actual behavior

I'm having an issue where CSS transitions do not work properly with LiveView when using `temporary_assigns`.
When rendering 5 items in a list, where each item can be highlighted by toggling a boolean field which toggling a class with a CSS transition.

* When I toggle the first 2 items the transition is working properly (animates the background). While the element is selected in the inspector, it updates the class on the selected element like expected.
* When I toggle all other (3) items the transition is not working (instantly changes background without animation). While the element is selected in the inspector, it seems to replace the whole element instead of updating it.

The CSS transition is probably just exposing an underlying issue, but it makes it easy to demonstrate.
Is there a logical explanation to this behaviour or could this be considered a bug?
FYI, when removing `temporary_assigns` and updating items in the list instead, all transitions work fine.

I've created a repo to demonstrate the issue:
https://github.com/sn3p/phoenix_live_temporary_assigns/

### Expected behavior

I would expect all item element to be updated and render the CSS transition when toggled. Or none at all when elements are replaced instead of updated, but not just the first 2 items.

**Comment by josevalim** [comment]
I may be wrong here but I recall you can’t add CSS classes to trigger animations? They need to either be there when the element is added or you need to trigger them especially. Otherwise it is not guaranteed to trigger depending on the event loop.

**Comment by josevalim** [comment]
Here is a related discussion: https://stackoverflow.com/questions/45275438/css-transition-is-not-trigger-by-add-classes-in-a-same-function

**Comment by chrismccord** [comment]
See the new `JS` interface for apply CSS transitions, which should fix your issue. As José said, the element will not transition if it already exists in the DOM, but your JS command API will allow you to apply the class exactly for this kind of usecase. For example, your template should be something like:

```
 <button phx-click={JS.transition("highlight") |> JS.push("highlight", value: %{id: id})}
```

I don't believe this is an issue on our side, but try that out and let us know if you're all set. Thanks for the concise example!

**Comment by sn3p** [comment]
> I may be wrong here but I recall you can’t add CSS classes to trigger animations? They need to either be there when the element is added or you need to trigger them especially. Otherwise it is not guaranteed to trigger depending on the event loop.

You're right, triggering CSS `animation` works a bit different. But triggering `transitions` should work when toggling classes, see this demo: https://codepen.io/snap/pen/mdMQvNw

The part I'm confused about is why toggling classes on the first 2 elements trigger the transition and for all other elements (3+) is doesn't trigger the transition. This suggest that the first 2 elements are updated and all other elements get replaced (which means there is nothing to transition from).

https://user-images.githubusercontent.com/935529/141979344-9bb4159b-d38b-4f57-8493-2069b024f6e5.mov

### Maybe I should rephrase my question
Why do the first 2 element get updated and other elements get replaces when using `temporary_assigns`? When updating elements without using `temporary_assigns` all elements are updated (not replaced) and the transitions work for all elements.

**Comment by sn3p** [comment]
> See the new `JS` interface for apply CSS transitions, which should fix your issue. As José said, the element will not transition if it already exists in the DOM, but your JS command API will allow you to apply the class exactly for this kind of usecase.

@chrismccord thanks I will look into this!

But I'm still confused by this behaviour and am looking for a logical explanation. You should be able to toggle class and trigger a `transition` (not `animation`) for existing elements that are already present in the DOM (see [this demo](https://codepen.io/snap/pen/mdMQvNw)).
The "problem" with `temporary_assigns` seem to be that the first 2 elements are updated (works) and all other elements are replaced (does not work). Without using `temporary_assigns` it works just fine.

I tried to rephrase my question with an example in the comment above.

**Comment by josevalim** [comment]
> This suggest that the first 2 elements are updated and all other elements get replaced (which means there is nothing to transition from).

How do you verify if they are being updated/replaced? Theoretically, if they have IDs, they should always be replaced, but sometimes this is a bit trickier than expected. It is one of the reasons why we want to get away from append/prepend in favor of something that better models collections.

**Comment by sn3p** [comment]
> How do you verify if they are being updated/replaced? Theoretically, if they have IDs, they should always be replaced, but sometimes this is a bit trickier than expected.

I assumed all elements except the first 2 were replaced instead of updated because:

* When one of the first 2 elements is selected in the browser inspector, you can see it updates the selected element class attribute when toggling.
* When one of the other elements (3+) is selected in the browser inspector, it looks like the selected element gets replaced as the `li` also flashes and the selection jumps to the next element.

https://user-images.githubusercontent.com/935529/141985536-81d9e70c-4d60-4d7c-b7d3-5d7554456a32.mov

But trying to verify this by comparing elements (before and after toggle) in the inspector, I see I was wrong as all elements are updated and not replaced. The comparison below shows that the 2nd (a) and 3rd (b) element are the same element before and after toggling.

![Screenshot 2021-11-16 at 13 55 56](https://user-images.githubusercontent.com/935529/141989087-7023ec2b-67bc-4991-9ff8-f8d0461aa8a2.png)

**Long story short:** I'm just puzzled why not all the elements show the same behaviour 😆 

@josevalim @chrismccord thank you for taking the time to look into this 🙏
